File: Sorting/merge sort/mergesort.py
from typing import List

def merge_sort(nums: List[int]) -> List[int]:
   
    n = len(nums)
    # Base case: A list with 0 or 1 element is already sorted.
    if n <= 1:
        return nums

    # 1. Divide: Python's slicing is very clean for this.
    mid = n // 2
    left_half = nums[:mid]
    right_half = nums[mid:]

    # 2. Conquer: Recursively sort both halves.
    sorted_left = merge_sort(left_half)
    sorted_right = merge_sort(right_half)

    # 3. Combine: Merge the sorted halves.
    return merge(sorted_left, sorted_right)

def merge(left: List[int], right: List[int]) -> List[int]:
   
    result = []
    i, j = 0, 0 # Pointers for left and right lists

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    
    # After the loop, one list may have remaining elements.
    # The .extend() method is a Pythonic way to add all remaining items.
    result.extend(left[i:])
    result.extend(right[j:])
    
    return result


def main():
    """
    Main function to handle I/O.
    """
    print("Enter numbers separated by spaces (e.g., 38 27 43 3 9 82 10):")
    line = input()

    try:
        numbers = [int(s) for s in line.split()]
    except ValueError:
        print("Invalid input.")
        return

    # Call the sorting function and store the new sorted list.
    sorted_numbers = merge_sort(numbers)

    print(f"Sorted Array: {sorted_numbers}")


if __name__ == "__main__":
    main()

File: Sorting/merge sort/mergesort.java
import java.util.*;

class solution{
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String line = sc.nextLine();

        String[] words = line.split(" ");
        int[] arr = new int[words.length];
        for(int i =0; i<arr.length; i++){
            arr[i] = Integer.parseInt(words[i]);
        } 

        mergeSort(arr);

        System.out.println("Sorted Array: " + Arrays.toString(sortedArr));
        sc.close();

    }


     public static int[] mergeSort(int[] nums) {
        int n = nums.length;
        // Base case for the recursion: an array with 0 or 1 elements is already sorted.
        if (n <= 1) {
            return nums;
        }

        // 1. Divide: Find the middle and create two subarrays.
        int mid = n / 2;
        int[] left = Arrays.copyOfRange(nums, 0, mid);
        int[] right = Arrays.copyOfRange(nums, mid, n);

        // 2. Conquer: Recursively sort both halves.
        int[] sortedLeft = mergeSort(left);
        int[] sortedRight = mergeSort(right);

        // 3. Combine: Merge the two sorted halves.
        return merge(sortedLeft, sortedRight);
    }

    
    private static int[] merge(int[] left, int[] right) {
        int[] result = new int[left.length + right.length];
        int i = 0; // Pointer for left array
        int j = 0; // Pointer for right array
        int k = 0; // Pointer for result array

        // Compare elements from left and right, adding the smaller one to the result.
        while (i < left.length && j < right.length) {
            if (left[i] <= right[j]) {
                result[k++] = left[i++];
            } else {
                result[k++] = right[j++];
            }
        }

        // After the main loop, one of the arrays might have leftover elements.
        // Copy the remaining elements from the left array (if any).
        while (i < left.length) {
            result[k++] = left[i++];
        }

        // Copy the remaining elements from the right array (if any).
        while (j < right.length) {
            result[k++] = right[j++];
        }

        return result;
    }
}
